gateway/ln-gateway/src/events.rs

```diff
@@ -22,4 +32,288 @@ pub const ALL_GATEWAY_EVENTS: [EventKind; 11] = [
+        .iter()
+        .filter(|e| {
+            if let Some((m, _)) = &e.module {
+                e.event_kind == event_kind && *m == module_kind
+            } else {
+                false
+            }
+        })
+        .collect::<Vec<_>>();
+    drop(event_kind);
+    drop(module_kind);
+    events
+}
+
+/// This function computes a "nested loop join" by first computing the cross
+/// product of the start event vector and the success/failure event vectors. The
+/// resulting cartesian product is then filtered according to the join predicate
+/// supplied in the parameters.
+///
+/// This function is intended for small data sets. If the data set relations
+/// grow, this function should implement a different join algorithm or be moved
+/// out of the gateway.
+pub(crate) fn join_events<Start, Success, Failure>(
+    start_events: &Vec<&PersistedLogEntry>,
+    success_events: &Vec<&PersistedLogEntry>,
+    failure_events: &Vec<&PersistedLogEntry>,
+    success_join_predicate: fn(Start, Success, u64) -> Option<(u64, Amount)>,
+    failure_join_predicate: fn(Start, Failure, u64) -> Option<u64>,
+) -> (Vec<(u64, Amount)>, Vec<u64>)
+where
+    Start: Event,
+    Success: Event,
+    Failure: Event,
+{
+    let cross_product = start_events
+        .iter()
+        .cartesian_product(success_events)
+        .collect::<Vec<_>>();
+    let success_stats = cross_product
+        .into_iter()
+        .filter_map(|(start, success)| {
+            if let Some(latency) = success.timestamp.checked_sub(start.timestamp) {
+                let start_event: Start =
+                    serde_json::from_value(start.value.clone()).expect("could not parse JSON");
+                let success_event: Success =
+                    serde_json::from_value(success.value.clone()).expect("could not parse JSON");
+                success_join_predicate(start_event, success_event, latency)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let cross_product = start_events
+        .iter()
+        .cartesian_product(failure_events)
+        .collect::<Vec<_>>();
+    let failure_stats = cross_product
+        .into_iter()
+        .filter_map(|(start, failure)| {
+            if let Some(latency) = failure.timestamp.checked_sub(start.timestamp) {
+                let start_event: Start =
+                    serde_json::from_value(start.value.clone()).expect("could not parse JSON");
+                let fail_event: Failure =
+                    serde_json::from_value(failure.value.clone()).expect("could not parse JSON");
+                failure_join_predicate(start_event, fail_event, latency)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    (success_stats, failure_stats)
+}
+
+/// Helper struct for storing computed data about outgoing and incoming
+/// payments.
+#[derive(Debug, Default)]
+pub struct StructuredPaymentEvents {
```

Comment:
Maybe this struct should have just `latencies`, `fees` and `latencies_failure`, so the caller can have two instances: outgoing and incoming? :shrug:

Reply:
Done